api/cluster/virtual_service.go

```diff
@@ -0,0 +1,188 @@
+package cluster
+
+import (
+	"context"
+	"encoding/json"
+	"fmt"
+	"net/url"
+	"strings"
+
+	istiov1beta1 "istio.io/api/networking/v1beta1"
+	v1beta1 "istio.io/client-go/pkg/apis/networking/v1beta1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+
+	"github.com/caraml-dev/merlin/models"
+	"github.com/caraml-dev/merlin/pkg/protocol"
+)
+
+const (
+	knativeIngressGateway          = "knative-serving/knative-ingress-gateway"
+	defaultIstioIngressGatewayHost = "istio-ingressgateway.istio-system.svc.cluster.local"
+)
+
+type VirtualService struct {
+	Name                    string
+	Namespace               string
+	VersionID               string
+	RevisionID              models.ID
+	Labels                  map[string]string
+	Protocol                protocol.Protocol
+	ModelVersionRevisionURL *url.URL
+}
+
+func NewVirtualService(modelService *models.Service, isvcURL string) (*VirtualService, error) {
+	modelVersionRevisionURL, err := url.Parse(isvcURL)
+	if err != nil {
+		return nil, fmt.Errorf("failed to parse model version revision url: %s", isvcURL)
+	}
+
+	if modelVersionRevisionURL.Scheme == "" {
+		veURL := "//" + isvcURL
+		modelVersionRevisionURL, err = url.Parse(veURL)
+		if err != nil {
+			return nil, fmt.Errorf("failed to parse model version revision url: %s", isvcURL)
+		}
+	}
+
+	return &VirtualService{
+		Name:                    fmt.Sprintf("%s-%s", modelService.ModelName, modelService.ModelVersion),
+		Namespace:               modelService.Namespace,
+		VersionID:               modelService.ModelVersion,
+		RevisionID:              modelService.RevisionID,
+		Labels:                  modelService.Metadata.ToLabel(),
+		Protocol:                modelService.Protocol,
+		ModelVersionRevisionURL: modelVersionRevisionURL,
+	}, nil
+}
+
+func (cfg VirtualService) BuildVirtualServiceSpec() (*v1beta1.VirtualService, error) {
+	modelVersionHost, err := cfg.getModelVersionHost()
+	if err != nil {
+		return nil, err
+	}
+
+	// We manually create model version revision to avoid Kserve's modification to transformer url if the name contains "transformer"
```

Comment:
I didn't understand this comment.. Which is the line that does "We manually create model version revision" ?

Reply:
I've removed this comment.

Previously, I faced an issue with model deployment for a model with the `transformer` word on its name. The inference service URL will replace the `transformer` name with an empty string.

For example, given a model version revision: `std-transformer-s-157-1`, the resulting URL will be `std-s-157-1-transformer.domain.com`

This is because on Kserve 0.11, they changed how to getServiceHost: https://github.com/kserve/kserve/commit/dbfcc97ea20ba990ed454d052dde9c9a7bc26eee#diff-0299d5ad4bb060d2faff81d5d3f2aedca8a9549c50254b1f0430d170822decefR82

With this PR, we prevent this issue via the new revision virtual service that routing to the resulted isvc URL. (E.g. the new revision virtual service host is `std-transformer-s-157-1.domain.com` which route to `std-s-157-1-transformer.domain.com`